poly_commit/src/hyrax/hyrax_impl.rs

```diff
@@ -0,0 +1,117 @@
+use arith::{ExtensionField, FieldSerde};
+use halo2curves::{ff::PrimeField, msm, CurveAffine};
+use polynomials::{
+    EqPolynomial, MultilinearExtension, MutRefMultiLinearPoly, MutableMultilinearExtension,
+    RefMultiLinearPoly,
+};
+
+use crate::hyrax::{
+    pedersen::{pedersen_commit, pedersen_setup},
+    PedersenParams,
+};
+
+pub(crate) fn hyrax_setup<C: CurveAffine + FieldSerde>(
+    local_vars: usize,
+    rng: impl rand::RngCore,
+) -> PedersenParams<C>
+where
+    C::Scalar: PrimeField,
+{
+    let pedersen_vars = (local_vars + 1) / 2;
+    let pedersen_length = 1 << pedersen_vars;
+
+    pedersen_setup(pedersen_length, rng)
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct HyraxCommitment<C: CurveAffine + FieldSerde>(pub Vec<C>);
+
+impl<C: CurveAffine + FieldSerde> FieldSerde for HyraxCommitment<C> {
+    const SERIALIZED_SIZE: usize = unimplemented!();
+
+    fn serialize_into<W: std::io::Write>(&self, writer: W) -> arith::FieldSerdeResult<()> {
+        self.0.serialize_into(writer)
+    }
+
+    fn deserialize_from<R: std::io::Read>(reader: R) -> arith::FieldSerdeResult<Self> {
+        let buffer: Vec<C> = Vec::deserialize_from(reader)?;
+        Ok(Self(buffer))
+    }
+}
+
+pub(crate) fn hyrax_commit<C: CurveAffine + FieldSerde>(
+    params: &PedersenParams<C>,
+    mle_poly: &impl MultilinearExtension<C::Scalar>,
+) -> HyraxCommitment<C>
+where
+    C::Scalar: ExtensionField + PrimeField,
+    C::ScalarExt: ExtensionField + PrimeField,
+{
+    let vars = mle_poly.num_vars();
+    let pedersen_vars = (vars + 1) / 2;
+    let pedersen_len = 1usize << pedersen_vars;
+    assert_eq!(pedersen_len, params.bases.len());
+
+    let commitments: Vec<C> = mle_poly
+        .hypercube_basis_ref()
+        .chunks(pedersen_len)
+        .map(|sub_hypercube| pedersen_commit(params, sub_hypercube))
+        .collect();
+
+    HyraxCommitment(commitments)
+}
+
+pub(crate) fn hyrax_open<C>(
+    params: &PedersenParams<C>,
+    mle_poly: &impl MultilinearExtension<C::Scalar>,
+    eval_point: &[C::Scalar],
+) -> (C::Scalar, Vec<C::Scalar>)
```

Comment:
nits: I think I'd prefer a struct `proof` or `opening` that wraps `Vec<C::Scalar>` like you did here
```
#[derive(Clone, Debug, Default)]
pub struct HyraxCommitment<C: CurveAffine + FieldSerde>(pub Vec<C>);
```


also add a comment that the first `C::Scalar` is the eval of the polynomial

Reply:
see https://github.com/PolyhedraZK/Expander/pull/200/commits/48b0df2dafa16dfac526d28057ad0b0827ca8444